### PR TITLE
parser: quote identifier with backquote when getting SQL digest

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Fixes https://github.com/pingcap/tidb/issues/14241

update parser and fix tests

### Related changes

- https://github.com/pingcap/parser/pull/1151

### Check List 

Tests 

- Unit test

### Release note 

- quote identifier with backquote when getting SQL digest

### CODE REVIEW

**Comment by morgo** [comment]
Please include a bind test or an integration test with a table + column that consists of a reserved word.

File: infoschema/tables_test.go [hunk 1]
```diff
@@ -1218,12 +1201,23 @@ func (s *testTableSuite) TestStmtSummaryHistoryTable(c *C) {
...
+	tk.MustQuery(sql).Check(testkit.Rows("Insert test test.test_summary <nil> 4 0 0 0 0 0 2 2 1 1 1 insert into test_summary values(1, 'a')"))
+
+	tk.MustExec("drop table if exists `table`")
+	tk.MustExec("create table `table`(`insert` int)")
+
+	sql = "select stmt_type, schema_name, table_names, index_names, exec_count, sum_cop_task_num, avg_total_keys," +
+		"max_total_keys, avg_processed_keys, max_processed_keys, avg_write_keys, max_write_keys, avg_prewrite_regions," +
+		"max_prewrite_regions, avg_affected_rows, query_sample_text " +
+		"from information_schema.statements_summary_history " +
+		"where digest_text like 'create table `table`%'"
```
Comment (xiongjiwei) [inline infoschema/tables_test.go:1219]: @morgo a test with a table + column that consists of a reserved word is added, PTAL

**Review by tangenta — COMMENTED** [review]
/LGTM

File: infoschema/tables_test.go (hunk 1 above)
Comment (morgo) [reply infoschema/tables_test.go:1219]: I meant a case like [this issue](https://github.com/pingcap/tidb/pull/22893). I had thought you were fixing this bug, but maybe not :-)

File: infoschema/tables_test.go [hunk 2]
```diff
@@ -1218,19 +1201,34 @@ func (s *testTableSuite) TestStmtSummaryHistoryTable(c *C) {
...
+	tk.MustExec("set global tidb_enable_stmt_summary = 1")
+	tk.MustExec("create table `table`(`insert` int)")
+	tk.MustExec("select `insert` from `table`")
+
+	sql = "select digest_text from information_schema.statements_summary_history;"
+	tk.MustQuery(sql).Check(testkit.Rows(
+		"select `insert` from `table`",
+		"create table `table` ( `insert` int )",
+		"set global `tidb_enable_stmt_summary` = ?",
+	))
```
Comment (xiongjiwei) [inline infoschema/tables_test.go:1231]: hi @morgo, is this test suitable for that issue?

Comment (morgo) [reply infoschema/tables_test.go:1231]: Yes. I had assumed (my mistake) that you were also fixing https://github.com/pingcap/tidb/issues/22388 - I have verified in your branch that it is not fixed yet though:

```sql
mysql> create binding for select * from t where a >=0 and a > 10 using select * from t ignore index(idx) where a >=0 and a > 10;
ERROR 1105 (HY000): line 1 column 23 near "-ada7-48ca-98cf-b424c3e140eb.t IGNORE INDEX (idx) WHERE a >= 0 AND a > 10" 
mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v4.0.0-beta.2-2175-gb0a9f5fef
Edition: Community
Git Commit Hash: b0a9f5fefcce52cfa6055f5ddae4ef5fdec8611f
Git Branch: pr/22893
UTC Build Time: 2021-02-25 02:40:16
GoVersion: go1.13.15
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.00 sec)
```

I think it is fine and we can disregard my previous comment. Issue #22388 can be solved separately.

Comment (xiongjiwei) [reply infoschema/tables_test.go:1231]: ok, I will take a look at this issue. BTW, can you please help me review this PR? 🤣

File: infoschema/tables_test.go [hunk 3]
```diff
@@ -1218,19 +1201,34 @@ func (s *testTableSuite) TestStmtSummaryHistoryTable(c *C) {
...
 
 	tk.MustExec("set global tidb_stmt_summary_history_size = 0")
 	tk.MustQuery(`select stmt_type, schema_name, table_names, index_names, exec_count, sum_cop_task_num, avg_total_keys,
 		max_total_keys, avg_processed_keys, max_processed_keys, avg_write_keys, max_write_keys, avg_prewrite_regions,
 		max_prewrite_regions, avg_affected_rows, query_sample_text, plan
 		from information_schema.statements_summary_history`,
 	).Check(testkit.Rows())
+
+	tk.MustExec("set global tidb_enable_stmt_summary = 0")
+	tk.MustExec("drop table if exists `table`")
```
Comment (xhebox) [inline infoschema/tables_test.go:1220]: I am wondering,  if the original SQL is not quoted but the query result, will that express the intention of test more clearer?

**Review by xhebox — COMMENTED** [review]
Rest LGTM

File: infoschema/tables_test.go [hunk 4]
```diff
@@ -957,12 +957,13 @@ func (s *testTableSuite) TestStmtSummaryTable(c *C) {
...
-		max_prewrite_regions, avg_affected_rows, query_sample_text
-		from information_schema.statements_summary
-		where digest_text like 'insert into t%'`,
-	).Check(testkit.Rows("Insert test test.t <nil> 4 0 0 0 0 0 2 2 1 1 1 insert into t values(1, 'a')"))
+
+	sql := "select stmt_type, schema_name, table_names, index_names, exec_count, sum_cop_task_num, avg_total_keys, " +
+		"max_total_keys, avg_processed_keys, max_processed_keys, avg_write_keys, max_write_keys, avg_prewrite_regions, " +
+		"max_prewrite_regions, avg_affected_rows, query_sample_text " +
+		"from information_schema.statements_summary " +
+		"where digest_text like 'insert into `t`%'"
```
[the remark this answers is not in the document]
Comment (xiongjiwei) [reply infoschema/tables_test.go:965]: the SQL contains backquotes, so I have to do this.

[the remark this answers is not in the document]
Comment (xhebox) [reply infoschema/tables_test.go:965]: OK. It can be solved by `strings.Builder` though...

File: infoschema/tables_test.go (hunk 3 above)
Comment (xiongjiwei) [reply infoschema/tables_test.go:1220]: `table` is a reserved word, must be quoted.

**Pull request merged**